### ISSUE TITLE
Segmentation fault when calling EVP_PKEY_CTX_get0_rsa_oaep_label with legacy key in 3.0.9

### ISSUE THREAD

**Issue by ljuzwiuk** [openssl/openssl#21335]
OpenSSL 3.0.9

We are utilizing an engine provided by a third party to retrieve keys from a Hardware Secure Module (HSM). The keys are in a legacy format. However, after updating from version 3.0.8 to 3.0.9, the following code with such a key is causing a segmentation fault:
```
unsigned char *oaep_label = NULL;
auto labellen = EVP_PKEY_CTX_get0_rsa_oaep_label(ctx, &oaep_label);
```

The line causing the issue is located within the `pkey_rsa_ctrl` function:
```
*(unsigned char **)p2 = rctx->oaep_label;
```
Both `p2` and `rctx->oaep_label` seem to be NULL. It appears that a simple NULL check is missing.

This problem surfaced after the introduction of the following [commit](https://github.com/openssl/openssl/pull/20780/files#diff-910515c87ac6c81bba3f87a197f149cdcc050330964f3ea26877217df4315e0eR2805).
```
ctx.ctrl_cmd = translation->ctrl_num;
```

**Comment by ljuzwiuk** [comment]
A bit more information is available. It turns out that simply having a legacy key is not enough to trigger the segmentation fault. A legacy context, which is created by the engine, is also required.

In such conditions:

- The `EVP_PKEY_CTX_set0_rsa_oaep_label` function does not work as expected. It returns success even though `pkey->data->oaep_label` remains NULL. You can see the relevant code [here](https://github.com/openssl/openssl/blob/de90e54bbe82e5be4fb9608b6f5c308bb837d355/crypto/rsa/rsa_pmeth.c#L573C1-L580C18). It appears that the `p1` parameter is never set and always equals 0, which affects the function's behavior.

- Similarly, the `EVP_PKEY_CTX_get0_rsa_oaep_label` function does not work correctly. As mentioned before, `p2` is a NULL pointer, leading to a null dereference. This can cause the segmentation fault. A simple NULL check for `p2` would fix the crash, but further work is required to make the function fully functional.

Stacktrace for sigsegv:
```
pkey_rsa_ctrl rsa_pmeth.c:587
evp_pkey_ctx_ctrl_int pmeth_lib.c:1330
EVP_PKEY_CTX_ctrl pmeth_lib.c:1364
evp_pkey_ctx_setget_params_to_ctrl ctrl_params_translate.c:2789
evp_pkey_ctx_get_params_to_ctrl ctrl_params_translate.c:2817
EVP_PKEY_CTX_get_params pmeth_lib.c:764
EVP_PKEY_CTX_get0_rsa_oaep_label rsa_lib.c:1132

```

**Comment by ljuzwiuk** [comment]
Here is a patch with possible fixes for `EVP_PKEY_CTX_set0_rsa_oaep_label` and `EVP_PKEY_CTX_get0_rsa_oaep_label`. It is important to note that due to the lack of comprehensive unit tests, it is uncertain whether these fixes will impact other functionalities.
```
diff --git a/crypto/evp/ctrl_params_translate.c b/crypto/evp/ctrl_params_translate.c
index ccafdfddd5..e5790734ce 100644
--- a/crypto/evp/ctrl_params_translate.c
+++ b/crypto/evp/ctrl_params_translate.c
@@ -636,8 +636,8 @@ static int default_fixup_args(enum state state,
                                                       ctx->p2, ctx->sz);
                 case OSSL_PARAM_OCTET_STRING:
                     return OSSL_PARAM_get_octet_string(ctx->params,
-                                                       ctx->p2, ctx->sz,
-                                                       &ctx->sz);
+                                                       &ctx->p2, ctx->sz,
+                                                       (size_t *)&ctx->p1);
                 case OSSL_PARAM_OCTET_PTR:
                     return OSSL_PARAM_get_octet_ptr(ctx->params,
                                                     ctx->p2, &ctx->sz);
@@ -695,6 +695,9 @@ static int default_fixup_args(enum state state,
                                    translation->param_data_type);
                     return 0;
                 }
+            } else if (state == PRE_PARAMS_TO_CTRL && ctx->action_type == GET) {
+                /* Assign output */
+                ctx->p2 = ctx->params->data;
             }
         }
         /* Any other combination is simply pass-through */
diff --git a/test/build.info b/test/build.info
index a5c1c65ddd..75846e05ac 100644
--- a/test/build.info
+++ b/test/build.info
@@ -41,8 +41,6 @@ IF[{- !$disabled{tests} -}]
           evp_pkey_provided_test evp_test evp_extra_test evp_extra_test2 \
           evp_fetch_prov_test evp_libctx_test ossl_store_test \
           v3nametest v3ext punycode_test \
-          evp_pkey_provided_test evp_test evp_extra_test evp_extra_test2 \
-          evp_fetch_prov_test v3nametest v3ext \
           crltest danetest bad_dtls_test lhash_test sparse_array_test \
           conf_include_test params_api_test params_conversion_test \
           constant_time_test verify_extra_test clienthellotest \
```

**Comment by t8m** [comment]
Would you like to create a pull request with the fix?

**Comment by ljuzwiuk** [comment]
Fixed in 3.0.10
